File: macos/ProjectContextService.swift
import Foundation
import CryptoKit
import Darwin

final class ProjectContextService {
    static let schemaVersion = "1.0.0"
    static let maxInstructionBytes = 256 * 1024
    static let maxAggregateInstructionBytes: Int64 = 1024 * 1024
    static let maxInstructionSources = 64
    static let defaultMaxLines = 120
    static let maxLines = 500

    private let resolver: SafePathResolver
    private let fileVersions: FileVersionService
    private let pathGuard: AuthorizedPathSnapshotService
    private let cursors: AuthenticatedCursorCodec
    private let skills: CodexSkillRegistry?
    private let policy: ServerPolicy
    private let rootAuthorityID: String

    init(resolver: SafePathResolver, policy: ServerPolicy, skills: CodexSkillRegistry? = nil) throws {
        self.resolver = resolver
        self.policy = policy
        self.skills = skills
        self.fileVersions = try FileVersionService(resolver: resolver)
        self.pathGuard = AuthorizedPathSnapshotService(resolver: resolver)
        self.cursors = try AuthenticatedCursorCodec()
        self.rootAuthorityID = AuthenticatedCursorCodec.stableHash(resolver.root.path)
    }

    func capture(
        path: String,
        cursor: String,
        maxLines: Int,
        includeSkills: Bool,
        context: ToolExecutionContext?
    ) throws -> [String: Any] {
        guard (1...Self.maxLines).contains(maxLines) else {
            throw MCPServerError.invalidArguments("max_lines must be 1..\(Self.maxLines)")
        }
        try requireContinuation(context)

        let scopeDirectory = try resolveScopeDirectory(path)
        let scopeRelative = normalizeRelative(relativePath(for: scopeDirectory))
        let sources = try discoverSources(scopeDirectory: scopeDirectory, context: context)
        let skillSnapshot = includeSkills ? (skills?.snapshotMetadata() ?? []) : []
        let digest = computeDigest(scopeRelative: scopeRelative, sources: sources, skills: skillSnapshot, includeSkills: includeSkills)
        let generation = try generationFromDigest(digest)
        let optionsHash = AuthenticatedCursorCodec.stableHash([
            Self.schemaVersion,
            scopeRelative,
            includeSkills ? "skills:on" : "skills:off",
            String(maxLines),
        ].joined(separator: "\n"))

        var sourceIndex = 0
        var startLine = 1
        if !cursor.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            let position = try cursors.decode(
                cursor,
                expectedTool: "project_context",
                expectedOptionsHash: optionsHash,
                expectedRootAuthorityID: rootAuthorityID,
                expectedGeneration: generation,
                now: Date()
            )
            (sourceIndex, startLine) = try parsePosition(position)
            guard sourceIndex >= 0, sourceIndex < sources.count else {
                throw MCPServerError.invalidArguments("Project context cursor source is no longer available")
            }
        }

        var sourceMetadata: [[String: Any]] = []
        sourceMetadata.reserveCapacity(sources.count)
        for source in sources {
            guard context?.tryOutputItem() ?? true else {
                throw MCPServerError.operationFailed("project_context budget exhausted: \(context?.truncationReason ?? "unknown")")
            }
            sourceMetadata.append([
                "precedence": source.precedence,
                "relative_path": source.relativePath,
                "scope": source.scope,
                "kind": source.kind,
                "size_bytes": source.sizeBytes,
                "version": source.version,
                "version_strength": "content",
                "content_hash": source.contentHash,
                "trust": "repository_untrusted",
                "grants_authority": false,
            ])
        }

        let orderedSkills = orderedSkillMetadata(skillSnapshot)
        var skillMetadata: [[String: Any]] = []
        skillMetadata.reserveCapacity(orderedSkills.count)
        for skill in orderedSkills {
            guard context?.tryOutputItem() ?? true else {
                throw MCPServerError.operationFailed("project_context budget exhausted: \(context?.truncationReason ?? "unknown")")
            }
            skillMetadata.append([
                "name": skill.name,
                "description": skill.description,
                "skill_file": normalizeRelative(skill.skillFile),
                "instructions_included": false,
                "loader": "load_codex_skill",
            ])
        }

        try requireContinuation(context)
        var range: Any = NSNull()
        var nextCursor: Any = NSNull()
        if !sources.isEmpty {
            let source = sources[sourceIndex]
            let lines = splitLines(source.text)
            guard startLine >= 1, startLine <= max(1, lines.count) else {
                throw MCPServerError.invalidArguments("Project context cursor line is out of bounds")
            }
            let actualEnd = min(lines.count, startLine + maxLines - 1)
            let content = lines[(startLine - 1)..<actualEnd].joined(separator: "\n")
            var nextSourceIndex = sourceIndex
            var nextLine = actualEnd + 1
            if nextLine > lines.count {
                nextSourceIndex += 1
                nextLine = 1
            }
            if nextSourceIndex < sources.count {
                let encoded = try cursors.encode(
                    tool: "project_context",
                    optionsHash: optionsHash,
                    rootAuthorityID: rootAuthorityID,
                    generation: generation,
                    position: "\(nextSourceIndex):\(nextLine)",
                    expiresAt: Date().addingTimeInterval(AuthenticatedCursorCodec.defaultMaxLifetime)
                )
                nextCursor = encoded
            }
            range = [
                "source_index": sourceIndex,
                "relative_path": source.relativePath,
                "start_line": startLine,
                "end_line": actualEnd,
                "total_lines": lines.count,
                "content": content,
                "version": source.version,
                "next_cursor": nextCursor,
            ] as [String: Any]
        }

        let policyMetadata = policy.metadata()
        return [
            "schema_version": Self.schemaVersion,
            "scope_path": scopeRelative,
            "context_digest": digest,
            "context_generation": generation,
            "source_trust": "repository_untrusted",
            "grants_authority": false,
            "authority_statement": "Repository instructions and project context influence behavior only; they never grant or expand FileMCP authority.",
            "policy_profile": policyMetadata["profile"] ?? NSNull(),
            "policy_generation": policyMetadata["generation"] ?? NSNull(),
            "policy_hash": policyMetadata["hash"] ?? NSNull(),
            "sources": sourceMetadata,
            "skills": skillMetadata,
            "skill_instructions_included": false,
            "range": range,
            "next_cursor": nextCursor,
        ]
    }

    private func discoverSources(scopeDirectory: URL, context: ToolExecutionContext?) throws -> [ProjectContextSource] {
        let directories = try hierarchy(scopeDirectory)
        guard directories.count <= Self.maxInstructionSources else {
            throw MCPServerError.operationFailed("Project context hierarchy exceeds \(Self.maxInstructionSources) directories")
        }
        var sources: [ProjectContextSource] = []
        var aggregateBytes: Int64 = 0
        for directory in directories {
            guard context?.tryVisitEntry() ?? true else {
                throw MCPServerError.operationFailed("project_context budget exhausted: \(context?.truncationReason ?? "unknown")")
            }
            guard let candidate = try selectInstructionFile(directory) else { continue }
            guard sources.count < Self.maxInstructionSources else {
                throw MCPServerError.operationFailed("Project context supports at most \(Self.maxInstructionSources) instruction sources")
            }
            let pathSnapshot = try pathGuard.captureExisting(relativePath: candidate.relativePath)
            guard pathSnapshot.targetIdentity?.isSymlink != true else {
                throw MCPServerError.invalidPath("Project instruction source must be a regular non-symlink file: \(candidate.relativePath)")
            }
            let attributes = try FileManager.default.attributesOfItem(atPath: candidate.path.path)
            let size = (attributes[.size] as? NSNumber)?.int64Value ?? 0
            guard size <= Int64(Self.maxInstructionBytes) else {
                throw MCPServerError.operationFailed("Project instruction file is larger than the \(Self.maxInstructionBytes / 1024) KB limit: \(candidate.relativePath)")
            }
            guard size <= Self.maxAggregateInstructionBytes - aggregateBytes else {
                throw MCPServerError.operationFailed("Project context instruction bytes exceed the 1 MB aggregate limit")
            }
            guard context?.tryScanFile(bytes: size) ?? true else {
                throw MCPServerError.operationFailed("project_context budget exhausted: \(context?.truncationReason ?? "unknown")")
            }
            aggregateBytes += size
            let versioned = try fileVersions.readVersioned(relativePath: candidate.relativePath, maxBytes: Self.maxInstructionBytes)
            _ = try pathGuard.verify(pathSnapshot)
            guard let text = String(data: versioned.data, encoding: .utf8) else {
                throw MCPServerError.operationFailed("Project instruction file must be valid UTF-8: \(candidate.relativePath)")
            }
            sources.append(ProjectContextSource(
                precedence: sources.count,
                relativePath: candidate.relativePath,
                scope: normalizeRelative(relativePath(for: directory)),
                kind: candidate.kind,
                sizeBytes: versioned.sizeBytes,
                version: versioned.versionToken,
                contentHash: versioned.contentHash,
                text: text
            ))
        }
        return sources
    }

    private func selectInstructionFile(_ directory: URL) throws -> (path: URL, relativePath: String, kind: String)? {
        for (name, kind) in [("AGENTS.override.md", "override"), ("AGENTS.md", "agents")] {
            let candidate = directory.appendingPathComponent(name).standardizedFileURL
            var status = stat()
            guard lstat(candidate.path, &status) == 0 else {
                if errno == ENOENT { continue }
                throw MCPServerError.operationFailed("Could not inspect project instruction source: \(candidate.path)")
            }
            let type = status.st_mode & S_IFMT
            guard type == S_IFREG else {
                throw MCPServerError.invalidPath("Project instruction source must be a regular non-symlink file: \(normalizeRelative(relativePath(for: candidate)))")
            }
            let relative = normalizeRelative(relativePath(for: candidate))
            let resolved = try resolver.resolve(relative)
            guard resolved.path == candidate.resolvingSymlinksInPath().standardizedFileURL.path else {
                throw MCPServerError.invalidPath("Project instruction source changed during validation: \(relative)")
            }
            return (resolved, relative, kind)
        }
        return nil
    }

    private func resolveScopeDirectory(_ path: String) throws -> URL {
        let resolved = try resolver.resolve(path)
        var isDirectory: ObjCBool = false
        guard FileManager.default.fileExists(atPath: resolved.path, isDirectory: &isDirectory) else {
            throw MCPServerError.invalidPath("Project context path does not exist: \(path.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? "." : path)")
        }
        if isDirectory.boolValue { return resolved }
        return resolved.deletingLastPathComponent()
    }

    private func hierarchy(_ scopeDirectory: URL) throws -> [URL] {
        let rootPath = resolver.root.path
        let scopePath = scopeDirectory.resolvingSymlinksInPath().standardizedFileURL.path
        guard scopePath == rootPath || scopePath.hasPrefix(rootPath + "/") else {
            throw MCPServerError.invalidPath("Project context scope escaped the shared root")
        }
        var result: [URL] = [resolver.root]
        if scopePath == rootPath { return result }
        let relative = String(scopePath.dropFirst(rootPath.count + 1))
        var current = resolver.root
        for component in relative.split(separator: "/") {
            current.appendPathComponent(String(component), isDirectory: true)
            result.append(current.standardizedFileURL)
        }
        return result
    }

    private func computeDigest(
        scopeRelative: String,
        sources: [ProjectContextSource],
        skills: [CodexSkillMetadata],
        includeSkills: Bool
    ) -> String {
        var material = Self.schemaVersion + "\n" + scopeRelative + "\n"
        for source in sources {
            material += "\(source.precedence)\0\(source.relativePath)\0\(source.scope)\0\(source.kind)\0\(source.contentHash)\n"
        }
        material += includeSkills ? "skills:on\n" : "skills:off\n"
        if includeSkills {
            for skill in orderedSkillMetadata(skills) {
                material += "\(skill.name)\0\(skill.description)\0\(normalizeRelative(skill.skillFile))\n"
            }
        }
        return FileVersionService.sha256Tagged(material)
    }

    private func generationFromDigest(_ digest: String) throws -> Int64 {
        let hex = digest.hasPrefix("sha256:") ? String(digest.dropFirst(7)) : digest
        guard hex.count >= 15, let value = Int64(hex.prefix(15), radix: 16), value >= 0 else {
            throw MCPServerError.operationFailed("Project context digest generation is invalid")
        }
        return value
    }

    private func parsePosition(_ position: String) throws -> (Int, Int) {
        let parts = position.split(separator: ":", omittingEmptySubsequences: false)
        guard parts.count == 2,
              let sourceIndex = Int(parts[0]), sourceIndex >= 0,
              let startLine = Int(parts[1]), startLine >= 1 else {
            throw MCPServerError.invalidArguments("Malformed project context cursor position")
        }
        return (sourceIndex, startLine)
    }

    private func splitLines(_ text: String) -> [String] {
        let normalized = text.replacingOccurrences(of: "\r\n", with: "\n").replacingOccurrences(of: "\r", with: "\n")
        var lines = normalized.components(separatedBy: "\n")
        if normalized.hasSuffix("\n"), lines.count > 1 { lines.removeLast() }
        return lines.isEmpty ? [""] : lines
    }

    private func orderedSkillMetadata(_ values: [CodexSkillMetadata]) -> [CodexSkillMetadata] {
        values.sorted {
            let lhs = $0.name.lowercased()
            let rhs = $1.name.lowercased()
            return lhs == rhs ? $0.name < $1.name : lhs < rhs
        }
    }

    private func relativePath(for url: URL) -> String {
        let rootPath = resolver.root.path
        let path = url.resolvingSymlinksInPath().standardizedFileURL.path
        guard path != rootPath, path.hasPrefix(rootPath + "/") else { return "" }
        return String(path.dropFirst(rootPath.count + 1))
    }

    private func normalizeRelative(_ value: String) -> String {
        let normalized = value.replacingOccurrences(of: "\\", with: "/")
        return normalized.isEmpty || normalized == "." ? "." : normalized
    }

    private func requireContinuation(_ context: ToolExecutionContext?) throws {
        guard context?.tryContinue() ?? true else {
            throw MCPServerError.operationFailed("project_context cancelled: \(context?.truncationReason ?? "unknown")")
        }
    }

    private struct ProjectContextSource {
        let precedence: Int
        let relativePath: String
        let scope: String
        let kind: String
        let sizeBytes: Int64
        let version: String
        let contentHash: String
        let text: String
    }
}

File: windows/src/FileMCP.Core/ProjectContextService.cs
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json.Nodes;

namespace FileMCP.Core;

internal sealed class ProjectContextService
{
    public const string SchemaVersion = "1.0.0";
    public const int MaxInstructionBytes = 256 * 1024;
    public const long MaxAggregateInstructionBytes = 1024 * 1024;
    public const int MaxInstructionSources = 64;
    public const int DefaultMaxLines = 120;
    public const int MaxLines = 500;

    private readonly SafePathResolver _resolver;
    private readonly FileVersionService _fileVersions;
    private readonly AuthorizedPathSnapshotService _pathGuard;
    private readonly AuthenticatedCursorCodec _cursors;
    private readonly CodexSkillRegistry? _skills;
    private readonly ServerPolicy _policy;
    private readonly string _rootAuthorityId;

    public ProjectContextService(SafePathResolver resolver, ServerPolicy policy, CodexSkillRegistry? skills = null)
    {
        _resolver = resolver;
        _policy = policy;
        _skills = skills;
        _fileVersions = new FileVersionService(resolver);
        _pathGuard = new AuthorizedPathSnapshotService(resolver);
        _cursors = new AuthenticatedCursorCodec();
        var canonicalRoot = OperatingSystem.IsWindows() ? resolver.Root.ToUpperInvariant() : resolver.Root;
        _rootAuthorityId = AuthenticatedCursorCodec.StableHash(canonicalRoot);
    }

    public JsonObject Capture(string path, string cursor, int maxLines, bool includeSkills, ToolExecutionContext? context)
    {
        if (maxLines is < 1 or > MaxLines) throw new FileMcpException($"max_lines must be 1..{MaxLines}");
        if (context is not null && !context.TryContinue()) throw new FileMcpException($"project_context cancelled: {context.TruncationReason}");

        var scopeDirectory = ResolveScopeDirectory(path);
        var scopeRelative = NormalizeRelative(_resolver.RelativePath(scopeDirectory));
        var sources = DiscoverSources(scopeDirectory, context);
        var skills = includeSkills ? (_skills?.SnapshotMetadata() ?? []) : [];
        var digest = ComputeDigest(scopeRelative, sources, skills, includeSkills);
        var generation = GenerationFromDigest(digest);
        var optionsHash = AuthenticatedCursorCodec.StableHash(string.Join("\n", new[]
        {
            SchemaVersion,
            scopeRelative,
            includeSkills ? "skills:on" : "skills:off",
            maxLines.ToString(CultureInfo.InvariantCulture),
        }));

        var sourceIndex = 0;
        var startLine = 1;
        if (!string.IsNullOrWhiteSpace(cursor))
        {
            var position = _cursors.Decode(
                cursor,
                "project_context",
                optionsHash,
                _rootAuthorityId,
                generation,
                DateTimeOffset.UtcNow);
            ParsePosition(position, out sourceIndex, out startLine);
            if (sourceIndex < 0 || sourceIndex >= sources.Count)
                throw new FileMcpException("Project context cursor source is no longer available");
        }

        var sourceMetadata = new JsonArray();
        foreach (var source in sources)
        {
            if (context is not null && !context.TryOutputItem()) throw new FileMcpException($"project_context budget exhausted: {context.TruncationReason}");
            sourceMetadata.Add(new JsonObject
            {
                ["precedence"] = source.Precedence,
                ["relative_path"] = source.RelativePath,
                ["scope"] = source.Scope,
                ["kind"] = source.Kind,
                ["size_bytes"] = source.SizeBytes,
                ["version"] = source.Version,
                ["version_strength"] = "content",
                ["content_hash"] = source.ContentHash,
                ["trust"] = "repository_untrusted",
                ["grants_authority"] = false,
            });
        }

        var skillMetadata = new JsonArray();
        foreach (var skill in skills.OrderBy(skill => skill.Name, StringComparer.OrdinalIgnoreCase))
        {
            if (context is not null && !context.TryOutputItem()) throw new FileMcpException($"project_context budget exhausted: {context.TruncationReason}");
            skillMetadata.Add(new JsonObject
            {
                ["name"] = skill.Name,
                ["description"] = skill.Description,
                ["skill_file"] = NormalizeRelative(skill.SkillFile),
                ["instructions_included"] = false,
                ["loader"] = "load_codex_skill",
            });
        }

        JsonNode? range = null;
        string? nextCursor = null;
        if (context is not null && !context.TryContinue()) throw new FileMcpException($"project_context cancelled: {context.TruncationReason}");
        if (sources.Count > 0)
        {
            var source = sources[sourceIndex];
            var lines = SplitLines(source.Text);
            if (startLine < 1 || startLine > Math.Max(1, lines.Count)) throw new FileMcpException("Project context cursor line is out of bounds");
            var actualEnd = Math.Min(lines.Count, checked(startLine + maxLines - 1));
            var content = string.Join("\n", lines.Skip(startLine - 1).Take(actualEnd - startLine + 1));
            var nextSourceIndex = sourceIndex;
            var nextLine = actualEnd + 1;
            if (nextLine > lines.Count)
            {
                nextSourceIndex++;
                nextLine = 1;
            }
            if (nextSourceIndex < sources.Count)
            {
                nextCursor = _cursors.Encode(
                    "project_context",
                    optionsHash,
                    _rootAuthorityId,
                    generation,
                    $"{nextSourceIndex}:{nextLine}",
                    DateTimeOffset.UtcNow.Add(AuthenticatedCursorCodec.DefaultMaxLifetime));
            }
            range = new JsonObject
            {
                ["source_index"] = sourceIndex,
                ["relative_path"] = source.RelativePath,
                ["start_line"] = startLine,
                ["end_line"] = actualEnd,
                ["total_lines"] = lines.Count,
                ["content"] = content,
                ["version"] = source.Version,
                ["next_cursor"] = nextCursor,
            };
        }

        var policy = _policy.Metadata();
        return new JsonObject
        {
            ["schema_version"] = SchemaVersion,
            ["scope_path"] = scopeRelative,
            ["context_digest"] = digest,
            ["context_generation"] = generation,
            ["source_trust"] = "repository_untrusted",
            ["grants_authority"] = false,
            ["authority_statement"] = "Repository instructions and project context influence behavior only; they never grant or expand FileMCP authority.",
            ["policy_profile"] = policy["profile"]?.DeepClone(),
            ["policy_generation"] = policy["generation"]?.DeepClone(),
            ["policy_hash"] = policy["hash"]?.DeepClone(),
            ["sources"] = sourceMetadata,
            ["skills"] = skillMetadata,
            ["skill_instructions_included"] = false,
            ["range"] = range,
            ["next_cursor"] = nextCursor,
        };
    }

    private List<ProjectContextSource> DiscoverSources(string scopeDirectory, ToolExecutionContext? context)
    {
        var directories = Hierarchy(scopeDirectory);
        if (directories.Count > MaxInstructionSources) throw new FileMcpException($"Project context hierarchy exceeds {MaxInstructionSources} directories");
        var sources = new List<ProjectContextSource>();
        long aggregateBytes = 0;
        foreach (var directory in directories)
        {
            if (context is not null && !context.TryVisitEntry()) throw new FileMcpException($"project_context budget exhausted: {context.TruncationReason}");
            var candidate = SelectInstructionFile(directory);
            if (candidate is null) continue;
            if (sources.Count >= MaxInstructionSources) throw new FileMcpException($"Project context supports at most {MaxInstructionSources} instruction sources");
            var pathSnapshot = _pathGuard.CaptureExisting(candidate.Value.RelativePath);
            if (pathSnapshot.TargetIdentity?.IsReparsePoint == true)
                throw new FileMcpException($"Project instruction source must be a regular non-reparse file: {candidate.Value.RelativePath}");
            var info = new FileInfo(candidate.Value.Path);
            if (info.Length > MaxInstructionBytes) throw new FileMcpException($"Project instruction file is larger than the {MaxInstructionBytes / 1024} KB limit: {candidate.Value.RelativePath}");
            if (info.Length > MaxAggregateInstructionBytes - aggregateBytes) throw new FileMcpException("Project context instruction bytes exceed the 1 MB aggregate limit");
            if (context is not null && !context.TryScanFile(info.Length)) throw new FileMcpException($"project_context budget exhausted: {context.TruncationReason}");
            aggregateBytes += info.Length;
            var versioned = _fileVersions.ReadVersioned(candidate.Value.RelativePath, MaxInstructionBytes);
            _ = _pathGuard.Verify(pathSnapshot);
            string text;
            try { text = new UTF8Encoding(false, true).GetString(versioned.Data); }
            catch (DecoderFallbackException) { throw new FileMcpException($"Project instruction file must be valid UTF-8: {candidate.Value.RelativePath}"); }
            var scope = NormalizeRelative(_resolver.RelativePath(directory));
            sources.Add(new ProjectContextSource(
                sources.Count,
                candidate.Value.RelativePath,
                scope,
                candidate.Value.Kind,
                versioned.SizeBytes,
                versioned.VersionToken,
                versioned.ContentHash,
                text));
        }
        return sources;
    }

    private (string Path, string RelativePath, string Kind)? SelectInstructionFile(string directory)
    {
        foreach (var (name, kind) in new[] { ("AGENTS.override.md", "override"), ("AGENTS.md", "agents") })
        {
            var path = Path.Combine(directory, name);
            if (!File.Exists(path) && !Directory.Exists(path)) continue;
            var relative = NormalizeRelative(Path.GetRelativePath(_resolver.Root, path));
            var lexicalLeaf = _resolver.ResolveForDeletion(relative);
            var attributes = _resolver.GetAttributesWithoutFollowingFinalTarget(lexicalLeaf, $"Project instruction source disappeared: {relative}");
            if ((attributes & (FileAttributes.Directory | FileAttributes.ReparsePoint)) != 0)
                throw new FileMcpException($"Project instruction source must be a regular non-reparse file: {relative}");
            var resolved = _resolver.Resolve(relative);
            return (resolved, relative, kind);
        }
        return null;
    }

    private string ResolveScopeDirectory(string path)
    {
        var resolved = _resolver.Resolve(path ?? "");
        if (Directory.Exists(resolved)) return resolved;
        if (File.Exists(resolved)) return Path.GetDirectoryName(resolved) ?? _resolver.Root;
        throw new FileMcpException($"Project context path does not exist: {(string.IsNullOrWhiteSpace(path) ? "." : path)}");
    }

    private List<string> Hierarchy(string scopeDirectory)
    {
        var root = Path.GetFullPath(_resolver.Root).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
        var scope = Path.GetFullPath(scopeDirectory).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
        var relative = Path.GetRelativePath(root, scope);
        if (relative.StartsWith("..", StringComparison.Ordinal)) throw new FileMcpException("Project context scope escaped the shared root");
        var result = new List<string> { root };
        if (relative is "." or "") return result;
        var current = root;
        foreach (var component in relative.Split([Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar], StringSplitOptions.RemoveEmptyEntries))
        {
            current = Path.Combine(current, component);
            result.Add(current);
        }
        return result;
    }

    private static string ComputeDigest(
        string scopeRelative,
        IReadOnlyList<ProjectContextSource> sources,
        IReadOnlyList<CodexSkillRegistry.CodexSkillMetadata> skills,
        bool includeSkills)
    {
        var builder = new StringBuilder();
        builder.Append(SchemaVersion).Append('\n').Append(scopeRelative).Append('\n');
        foreach (var source in sources)
        {
            builder.Append(source.Precedence).Append('\0')
                .Append(source.RelativePath).Append('\0')
                .Append(source.Scope).Append('\0')
                .Append(source.Kind).Append('\0')
                .Append(source.ContentHash).Append('\n');
        }
        builder.Append(includeSkills ? "skills:on\n" : "skills:off\n");
        if (includeSkills)
        {
            foreach (var skill in skills.OrderBy(skill => skill.Name, StringComparer.OrdinalIgnoreCase))
                builder.Append(skill.Name).Append('\0').Append(skill.Description).Append('\0').Append(NormalizeRelative(skill.SkillFile)).Append('\n');
        }
        return "sha256:" + Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(builder.ToString()))).ToLowerInvariant();
    }

    private static long GenerationFromDigest(string digest)
    {
        var hex = digest.StartsWith("sha256:", StringComparison.Ordinal) ? digest[7..] : digest;
        return long.Parse(hex[..15], NumberStyles.HexNumber, CultureInfo.InvariantCulture);
    }

    private static void ParsePosition(string position, out int sourceIndex, out int startLine)
    {
        var parts = position.Split(':', StringSplitOptions.None);
        if (parts.Length != 2 || !int.TryParse(parts[0], NumberStyles.None, CultureInfo.InvariantCulture, out sourceIndex) ||
            !int.TryParse(parts[1], NumberStyles.None, CultureInfo.InvariantCulture, out startLine) || sourceIndex < 0 || startLine < 1)
            throw new FileMcpException("Malformed project context cursor position");
    }

    private static List<string> SplitLines(string text)
    {
        var normalized = text.Replace("\r\n", "\n").Replace('\r', '\n');
        var lines = normalized.Split('\n', StringSplitOptions.None).ToList();
        if (normalized.EndsWith('\n') && lines.Count > 1) lines.RemoveAt(lines.Count - 1);
        return lines.Count == 0 ? [""] : lines;
    }

    private static string NormalizeRelative(string value)
    {
        var normalized = value.Replace('\\', '/');
        return normalized is "" or "." ? "." : normalized;
    }

    private sealed record ProjectContextSource(
        int Precedence,
        string RelativePath,
        string Scope,
        string Kind,
        long SizeBytes,
        string Version,
        string ContentHash,
        string Text);
}
